store/etcd/put.go

```diff
@@ -3,236 +3,156 @@ package etcd
-			entry.Version = pbrse.Version
-			entry.Ruleset = rulesetFromProtobuf(pbrse.Ruleset)
-
-			return store.ErrNotModified
-		}
-
-		// make sure signature didn't change
-		rawSig := stm.Get(s.signaturesPath(path))
-		if rawSig != "" {
-			var pbsig pb.Signature
-
-			err := proto.Unmarshal([]byte(rawSig), &pbsig)
-			if err != nil {
-				s.Logger.Debug().Err(err).Str("signature", rawSig).Msg("put: signature unmarshalling failed")
-				return errors.Wrap(err, "failed to decode ruleset signature")
-			}
-
-			err = compareSignature(signatureFromProtobuf(&pbsig), sig)
-			if err != nil {
-				return err
-			}
-		}
-
-		// if no signature found, create one
-		if rawSig == "" {
-			b, err := proto.Marshal(signatureToProtobuf(sig))
-			if err != nil {
-				return errors.Wrap(err, "failed to encode updated signature")
-			}
-
-			stm.Put(s.signaturesPath(path), string(b))
-		}
+		p := rulesPutter{s, stm}
+		entry, err = p.put(ctx, path, ruleset)
+		return err
+	}
 
-		// update checksum
-		stm.Put(s.checksumsPath(path), checksum)
+	_, err = concurrency.NewSTM(s.Client, txfn, concurrency.WithAbortContext(ctx))
+	if err != nil && err != store.ErrRulesetNotModified && !store.IsValidationError(err) {
+		return nil, errors.Wrap(err, "failed to put ruleset")
+	}
 
-		// create a new ruleset version
-		k, err := ksuid.NewRandom()
-		if err != nil {
-			return errors.Wrap(err, "failed to generate ruleset version")
-		}
-		version := k.String()
+	return entry, err
+}
 
-		err = s.putVersions(stm, path, version)
-		if err != nil {
-			return err
-		}
+// rulesPutter is responsible for validating and storing rules, updating checksums and other actions
+// that are required in order to add a new ruleset version correctly.
+type rulesPutter struct {
+	s   *RulesetService
+	stm concurrency.STM
+}
 
-		re := store.RulesetEntry{
-			Path:    path,
-			Version: version,
-			Ruleset: ruleset,
-		}
+func (p *rulesPutter) put(ctx context.Context, path string, ruleset *regula.Ruleset) (*store.RulesetEntry, error) {
+	var err error
 
-		err = s.putEntry(stm, &re)
-		if err != nil {
-			return err
-		}
+	entry := store.RulesetEntry{
+		Path:    path,
+		Ruleset: ruleset,
+	}
 
-		// update the pointer to the latest ruleset
-		stm.Put(s.latestRulesetPath(path), s.entriesPath(path, version))
+	// validate the ruleset
+	entry.Signature, err = p.validateRuleset(p.stm, path, ruleset)
+	if err != nil {
+		return nil, err
+	}
 
-		entry = re
-		return nil
+	// encode rules
+	data, err := proto.Marshal(rulesetToProtobuf(ruleset))
+	if err != nil {
+		return nil, err
 	}
 
-	_, err = concurrency.NewSTM(s.Client, txfn, concurrency.WithAbortContext(ctx))
-	if err != nil && err != store.ErrNotModified && !store.IsValidationError(err) {
-		return nil, errors.Wrap(err, "failed to put ruleset")
+	// update checksum if rules have changed
+	changed, err := p.updateChecksum(p.stm, path, data)
+	if err != nil {
+		return nil, err
 	}
 
-	return &entry, err
-}
+	if !changed {
```

Comment:
Maybe we should do this firstly, that case, if the ruleset hasn't changed, we return it directly without checking if the ruleset satisfies the signature.